File: AtCoder/AtCoder_Improvement_Guidelines/2-3/D - Grid Repainting.py
h,w=map(int,input().split())
s=[]; b=0
for i in [0]*h:
    r=input()
    b+=r.count("#")
    s+=[list(r)]
s[0][0]=1
q=[(0,0)]
while q:
    y,x=q.pop(0)
    for nx,ny in (x,y-1),(x+1,y),(x,y+1),(x-1,y):
        if h>ny>=0<=nx<w and s[ny][nx]==".":
            s[ny][nx]=s[y][x]+1
            q+=[(ny,nx)]
print(-1 if s[h-1][w-1]=="." else h*w-b-s[h-1][w-1])

File: Archive/AtCoder/AtCoder_Improvement_Guidelines/2-3/D - Grid Repainting.cpp
#include<bits/stdc++.h>
using namespace std;
int main(){
  int H,W; cin>>H>>W;
  string s[H]; int b=0;
  for(int i=0;i<H;i++){
    cin>>s[i]; b+=count(s[i].begin(),s[i].end(),'#');
  }  
  int p[H][W]={}; p[0][0]=1;
  queue<pair<int,int>>q; q.push({0,0});
  while(q.size()){
    auto i=q.front(); q.pop();
    int y=i.first,x=i.second;
    for(auto d:vector<pair<int,int>>{{1,0},{0,1},{-1,0},{0,-1}}){
      int ny=y+d.first,nx=x+d.second;
      if(0<=ny&&ny<H&&0<=nx&&nx<W&&s[ny][nx]=='.'&&!p[ny][nx]){
        p[ny][nx]=p[y][x]+1; q.push({ny,nx});
      }
    }
  }
  cout<<(p[H-1][W-1]?H*W-p[H-1][W-1]-b:-1)<<endl;
}